src/disco/shred/test_shredder.c

```diff
@@ -141,6 +141,85 @@ test_shredder_pcap( void ) {
 #endif /* FD_HAS_HOSTED */
 
 
+static void
+test_multiple_shredders( void ) {
+  fd_rng_t _rng[ 1 ]; fd_rng_t * r = fd_rng_join( fd_rng_new( _rng, 0U, 0UL ) );
+
+  signer_ctx_t signer_ctx[ 1 ];
+  signer_ctx_init( signer_ctx, test_private_key );
+
+  #define SHREDDERS 4
+
+  fd_shredder_t   _shredders[ SHREDDERS ];
+  fd_shredder_t *  shredders[ SHREDDERS ];
+  /* Initialize all the shredders */
+  for( ulong i=0; i<SHREDDERS; i++ ) {
+    FD_TEST( &_shredders[ i ]==fd_shredder_new( &_shredders[ i ], test_signer, signer_ctx, (ushort)0 ) );
+    shredders[ i ] = fd_shredder_join( &_shredders[ i ] );
+    FD_TEST( shredders[ i ] );
+  }
+
+  fd_entry_batch_meta_t meta[ 1 ];
+  fd_memset( meta, 0, sizeof( fd_entry_batch_meta_t ) );
+
+  fd_fec_set_t _set[ 1 ];
+  for( ulong j=0UL; j<FD_REEDSOL_DATA_SHREDS_MAX;   j++ ) _set->data_shreds[   j ] = fec_set_memory_1 + 2048UL*j;
+  for( ulong j=0UL; j<FD_REEDSOL_PARITY_SHREDS_MAX; j++ ) _set->parity_shreds[ j ] = fec_set_memory_2 + 2048UL*j;
+
+  ulong data_shred_cnt   = 0;
+  ulong parity_shred_cnt = 0;
+  uchar data[ 1024UL*1024UL ];
+  ulong idx  = 0UL;
+  ulong slot = 1UL;
+
+  /* Randomly choose a shredder to process a batch until the data buffer is exhausted. */
+  while( idx<1024UL*1024UL ) {
+    ulong sz        = ( fd_rng_ulong_roll( r, 10UL )+1UL )*1000UL;
+    ulong batch_sz  = fd_ulong_if( idx+sz>1024UL*1024UL, 1024UL*1024UL-idx, sz );
+    ulong shredder  = fd_rng_ulong_roll( r, SHREDDERS );
+    for( ulong i=0; i<SHREDDERS; i++ ) {
+      if( FD_UNLIKELY( i==shredder ) ) {
+        FD_TEST( fd_shredder_init_batch( shredders[ i ], data+idx, batch_sz, slot, meta ) );
+        ulong fec_sets = fd_shredder_count_fec_sets( batch_sz );
+        for( ulong j=0; j<fec_sets; j++ ) {
+          FD_TEST( fd_shredder_next_fec_set( shredders[ i ], _set ) );
+          data_shred_cnt   += _set->data_shred_cnt;
+          parity_shred_cnt += _set->parity_shred_cnt;
+        }
+        FD_TEST( fd_shredder_fini_batch( shredders[ i ] ) );
+      } else {
+        FD_TEST( fd_shredder_skip_batch( shredders[ i ], batch_sz, slot ));
+      }
+    }
+    for( ulong i=0; i<SHREDDERS; i++ ) {
+      FD_TEST( shredders[ i ]->data_idx_offset==data_shred_cnt );
+      FD_TEST( shredders[ i ]->parity_idx_offset==parity_shred_cnt );
+    }
+    idx  += batch_sz;
+    /* Increment the slot every 100_000 bytes. */
+    if( FD_UNLIKELY( idx/100000UL>=slot ) ) {
+      slot++;
+      data_shred_cnt   = 0;
+      parity_shred_cnt = 0;
+    }
+  }
+
+  /* Make all the shredders process the same data and compare the outputs. */
+  fd_fec_set_t _sets[ SHREDDERS ];
+  for( ulong i=0; i<SHREDDERS; i++ ) {
+    for( ulong j=0UL; j<FD_REEDSOL_DATA_SHREDS_MAX;   j++ ) ( &_sets[ i ] )->data_shreds[   j ] = fec_set_memory_1 + 2048UL*j;
```

Comment:
```suggestion
    for( ulong j=0UL; j<FD_REEDSOL_DATA_SHREDS_MAX;   j++ ) sets[ i ].data_shreds[   j ] = fec_set_memory_1 + 2048UL*j;
```

Reply:
But actually, the problem is deeper than just formatting. Each FEC set is pointing to the same memory, which means they're all going to overwrite each other, and this comparison test wouldn't actually test anything

Reply:
But in general, if you want a pointer to the ith element, `(sets+i)` is your friend. If you're going to dereference it immediately with `->`, I'd just go with `sets[i]`.